pkg/controller/queuejob/queuejob_controller_ex.go

```diff
@@ -1105,342 +1107,354 @@ func (qjm *XController) ScheduleNext() {
+						klog.V(4).Infof("[ScheduleNext] [Agent Mode] HOL quota evaluation not enabled for '%s/%s' at %s activeQ=%t Unsched=%t &qj=%p Version=%s Status=%+v", qj.Namespace,
+							qj.Name, time.Now().Sub(HOLStartTime), qjm.qjqueue.IfExistActiveQ(qj), qjm.qjqueue.IfExistUnschedulableQ(qj), qj, qj.ResourceVersion, qj.Status)
 					}
-				} else {
-					klog.V(4).Infof("[ScheduleNext] HOL quota evaluation not enabled for %s at %s activeQ=%t Unsched=%t &qj=%p Version=%s Status=%+v", qj.Name, time.Now().Sub(HOLStartTime), qjm.qjqueue.IfExistActiveQ(qj), qjm.qjqueue.IfExistUnschedulableQ(qj), qj, qj.ResourceVersion, qj.Status)
-				}
 
-				// If quota evalauation sucedeed or quota evaluation not enabled set the appwrapper to be dispatched
-				if fits {
-					// aw is ready to go!
-					apiQueueJob, e := qjm.queueJobLister.AppWrappers(qj.Namespace).Get(qj.Name)
-					// apiQueueJob's ControllerFirstTimestamp is only microsecond level instead of nanosecond level
-					if e != nil {
-						klog.Errorf("[ScheduleNext] Unable to get AW %s from API cache &aw=%p Version=%s Status=%+v err=%#v", qj.Name, qj, qj.ResourceVersion, qj.Status, err)
-						if qjm.quotaManager != nil && quotaFits {
-							//Quota was allocated for this appwrapper, release it.
-							qjm.quotaManager.Release(qj)
+					// If quota evalauation sucedeed or quota evaluation not enabled set the appwrapper to be dispatched
+					if fits {
+						// aw is ready to go!
+						tempAW, retryErr := qjm.getAppWrapper(qj.Namespace, qj.Name, "[ScheduleNext] [Agent Mode]  -- ready to dispatch")
+						if retryErr != nil {
+							klog.Errorf("[ScheduleNext] [Agent Mode] Failed to get fresh copy of the app wrapper '%s/%s' to update status, err = %v", qj.Namespace, qj.Name, err)
+							return retryErr
 						}
-						return
-					}
-					// make sure qj has the latest information
-					if larger(apiQueueJob.ResourceVersion, qj.ResourceVersion) {
-						klog.V(4).Infof("[ScheduleNext] %s found more recent copy from cache          &qj=%p          qj=%+v", qj.Name, qj, qj)
-						klog.V(4).Infof("[ScheduleNext] %s found more recent copy from cache &apiQueueJob=%p apiQueueJob=%+v", apiQueueJob.Name, apiQueueJob, apiQueueJob)
-						apiQueueJob.DeepCopyInto(qj)
-					}
-					desired := int32(0)
-					for i, ar := range qj.Spec.AggrResources.Items {
-						desired += ar.Replicas
-						qj.Spec.AggrResources.Items[i].AllocatedReplicas = ar.Replicas
-					}
-					qj.Status.CanRun = true
-					qj.Status.FilterIgnore = true // update CanRun & Spec.  no need to trigger event
-					// Handle k8s watch race condition
-					if err := qjm.updateEtcd(qj, "ScheduleNext - setCanRun"); err == nil {
+						desired := int32(0)
+						for i, ar := range tempAW.Spec.AggrResources.Items {
+							desired += ar.Replicas
+							tempAW.Spec.AggrResources.Items[i].AllocatedReplicas = ar.Replicas
```

Comment:
@dmatch01 and @asm582  do you know what this block of code does ? Why is it necessary ?

Reply:
I think this dead code, Items are not used